### PR TITLE
黎声第二次作业

### PR DESCRIPTION
**Pull request by gobuddy**

### CODE REVIEW

**Comment by gobuddy** [comment]
@kaichaosun

File: projects/lesson-2/runtime/src/substratekitties.rs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+        AllKittiesArray get(kitty_by_index): map u64 => T::Hash;
+        AllKittiesCount get(all_kitties_count): u64;
+        AllKittiesIndex: map T::Hash => u64;
+
+        OwnedKittiesArray get(kitty_of_owner_by_index): map (T::AccountId, u64) => T::Hash;
+        OwnedKittiesCount get(owned_kitty_count): map T::AccountId => u64;
+        OwnedKittiesIndex: map T::Hash => u64;
+
+        Nonce: u128;
```
Comment (kaichaosun) [inline projects/lesson-2/runtime/src/substratekitties.rs]: u64 should be enough for our usage. Check the [max value](https://doc.rust-lang.org/std/primitive.u64.html#examples-1)

File: projects/lesson-3/runtime/src/kitties.rs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		/// Create a new kitty
+		pub fn create(origin) {
+			let sender = ensure_signed(origin)?;
+			let count = Self::kitties_count();
+			if count == u32::max_value() {
+				return Err("Kitties count overflow");
+			}
+			let payload = (<system::Module<T>>::random_seed(), sender, <system::Module<T>>::extrinsic_index(), <system::Module<T>>::block_number());
+			let dna = payload.using_encoded(blake2_128);
```
Comment (kaichaosun) [inline projects/lesson-3/runtime/src/kitties.rs]: Good one here. 👍

File: projects/lesson-3/runtime/src/kitties.rs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+#[derive(Encode, Decode, Default)]
+pub struct Kitty(pub [u8; 16]);
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Kitties {
+		/// Stores all the kitties, key is the kitty id / index
+		pub Kitties get(kitty): map u32 => Option<Kitty>;
+		/// Stores the total number of kitties. i.e. the next kitty index
+		pub KittiesCount get(kitties_count): u32;
```
Comment (kaichaosun) [inline projects/lesson-3/runtime/src/kitties.rs]: You need storage to track kitty owner logic.

File: projects/lesson-3/runtime/src/kitties.rs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+			let sender = ensure_signed(origin)?;
+			let count = Self::kitties_count();
+			if count == u32::max_value() {
+				return Err("Kitties count overflow");
+			}
+			let payload = (<system::Module<T>>::random_seed(), sender, <system::Module<T>>::extrinsic_index(), <system::Module<T>>::block_number());
+			let dna = payload.using_encoded(blake2_128);
+			let kitty = Kitty(dna);
+			Kitties::insert(count, kitty);
+			KittiesCount::put(count + 1);
```
Comment (kaichaosun) [inline projects/lesson-3/runtime/src/kitties.rs]: Return `Ok(())` at end of the funciton.

**Pull request closed without merging**